File: 2492. Minimum Score of a Path Between Two Cities/Solution.py
class Solution:
    def minScore(self, n: int, roads: List[List[int]]) -> int:
        uf = UFDS(n)
        for x, y, w in roads:
            uf.union(x, y, w)
        return uf.getDist(1)

class UFDS:
    def __init__(self, n):
        self.parent = [i for i in range(n + 1)]
        self.rank = [0 for i in range(n + 1)]
        self.dist = [10001 for i in range(n + 1)]

    def find(self, x):
        if self.parent[x] != x: self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def same(self, x, y):
        return self.find(x) == self.find(y)
    
    def union(self, x, y, w):
        a, b = self.find(x), self.find(y)
        self.dist[a] = min(self.dist[a], self.dist[b], w)
        self.dist[b] = min(self.dist[a], self.dist[b], w)
        if self.same(x, y): return
        if self.rank[a] < self.rank[b]:
            self.parent[a] = b
        else:
            self.parent[b] = a
            if self.rank[a] == self.rank[b]: self.rank[a] += 1

    def getDist(self, x):
        return self.dist[self.find(x)]

File: 2492. Minimum Score of a Path Between Two Cities/Solution.java
class Solution {
    public int minScore(int n, int[][] roads) {
        UFDS uf = new UFDS(n);
        for (int[] r : roads) {
            uf.union(r[0], r[1], r[2]);
        }
        return uf.getDist(1);
    }
}

class UFDS {
    int[] parent, rank, dist;

    public UFDS(int n) {
        parent = new int[n + 1];
        rank = new int[n + 1];
        dist = new int[n + 1];
        for (int i = 0; i < n + 1; i++) {
            parent[i] = i;
            dist[i] = 10001;
        }
    }

    public int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    public boolean same(int x, int y) {
        return find(x) == find(y);
    }

    public void union(int x, int y, int w) {
        int a = find(x), b = find(y);
        dist[b] = Math.min(Math.min(dist[b], dist[a]), w);
        dist[a] = Math.min(Math.min(dist[a], dist[b]), w);
        if (same(x, y)) return;
        if (rank[a] < rank[b]) {
            parent[a] = b;
        } else {
            parent[b] = a;
            if (rank[a] == rank[b]) rank[a]++;
        }
    }

    public int getDist(int x) {
        return dist[find(x)];
    }
}